### PR TITLE
Fix(MSSQL): fix results fetching

### PR DESCRIPTION
**Pull request by treysp**
Background:
- pymssql automatically opens transactions, causing hung sessions on some CREATE statements
- #1532 fixed this by wrapping all `execute` calls in a transaction to ensure transactions would be closed

Current issue:
- The transaction wrapping in #1532 closes the pymssql transaction after executing the query, such that query results are not available to be fetched by the `fetchone` or `fetchall` methods

This PR:
- Creates a new MSSQL method `execute_and_fetch` that both executes a query and fetches its results. The method copies the base engine adapter `execute` code because we must fetch the results from inside the transaction context manager.
- Overrides the base engine adapter `fetchone` and `fetchall` methods.

### CODE REVIEW

File: sqlmesh/core/engine_adapter/mssql.py [hunk 1]
```diff
@@ -178,7 +180,7 @@ def _get_data_objects(
         Returns all the data objects that exist in the given schema and catalog.
         """
         if not catalog_name:
-            catalog_name = self.fetchone("select DB_NAME()")[0]
+            catalog_name = self.fetchone("select DB_NAME()", quote_identifiers=True)[0]
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/mssql.py:183]: why are you doing quote_identifiers=True here?

File: sqlmesh/core/engine_adapter/mssql.py [hunk 2]
```diff
@@ -188,15 +190,15 @@ def _get_data_objects(
             FROM {catalog_name}.INFORMATION_SCHEMA.TABLES
             WHERE TABLE_SCHEMA LIKE '%{schema_name}%'
         """
-        dataframe: pd.DataFrame = self.fetchdf(query)
+        results = self.fetchall(query, quote_identifiers=True)
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/mssql.py:193]: same here, if you care just quote your query?

File: sqlmesh/core/engine_adapter/mssql.py [hunk 3]
```diff
@@ -206,3 +208,80 @@ def _truncate_table(self, table_name: TableName) -> str:
     def _to_sql(self, expression: exp.Expression, quote: bool = True, **kwargs: t.Any) -> str:
         sql = super()._to_sql(expression, quote=quote, **kwargs)
         return f"{sql};"
+
+    @overload
+    def execute_and_fetch(
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/mssql.py:214]: why do these exists in the public space? i don't think this is necessary

File: sqlmesh/core/engine_adapter/mssql.py [hunk 4]
```diff
@@ -206,3 +209,80 @@ def _truncate_table(self, table_name: TableName) -> str:
...
+        self,
+        expression: t.Union[str, exp.Expression],
+        fetch_one: Literal[True],
+        ignore_unsupported_errors: bool = False,
+        quote_identifiers: bool = True,
+        **kwargs: t.Any,
+    ) -> t.Tuple:
+        ...
```
Comment (GeorgeSittas) [inline sqlmesh/core/engine_adapter/mssql.py:233]: I think we may be able to reduce the visual noise here by doing something similar to this: https://github.com/tobymao/sqlglot/blob/main/sqlglot/__init__.py#L89

File: sqlmesh/core/engine_adapter/mssql.py [hunk 5]
```diff
@@ -206,3 +208,80 @@ def _truncate_table(self, table_name: TableName) -> str:
...
+
+    def execute_and_fetch(
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/mssql.py:235]: move this into fetchall, and then have fetch_one just return 1

Comment (tobymao) [reply sqlmesh/core/engine_adapter/mssql.py:235]: from fetchall

Comment (treysp) [reply sqlmesh/core/engine_adapter/mssql.py:235]: So fetchone should call fetchall and return the first entry?

Comment (tobymao) [reply sqlmesh/core/engine_adapter/mssql.py:235]: yes, and check the behavior on how it should raise

Comment (tobymao) [reply sqlmesh/core/engine_adapter/mssql.py:235]: actually i change my mind

Comment (tobymao) [reply sqlmesh/core/engine_adapter/mssql.py:235]: can't you just do 

with self.transaction():
   self.execute()
   return self.cursor.fetchone()

File: sqlmesh/core/engine_adapter/mssql.py [hunk 6]
```diff
@@ -188,15 +191,15 @@ def _get_data_objects(
             FROM {catalog_name}.INFORMATION_SCHEMA.TABLES
             WHERE TABLE_SCHEMA LIKE '%{schema_name}%'
         """
-        dataframe: pd.DataFrame = self.fetchdf(query)
```
Comment (tobymao) [inline sqlmesh/core/engine_adapter/mssql.py:191]: why did you get rid of this? why don't you just fix fetchdf with a transaction wrap?

**Pull request closed without merging**